src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -901,40 +937,99 @@ private async Task<CompilationCompletedEvent> ProcessCompilationEventsCoreAsync(
             }
         }
 
-        private void ProcessEvent(CompilationEvent e, AnalysisScope analysisScope, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
+        private async Task ProcessEventAsync(CompilationEvent e, AnalysisScope analysisScope, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
         {
-            if (TryProcessEventCore(e, analysisScope, analysisStateOpt, cancellationToken))
+            EventProcessedState eventProcessedState = await TryProcessEventCoreAsync(e, analysisScope, analysisStateOpt, cancellationToken).ConfigureAwait(false);
+
+            ImmutableArray<DiagnosticAnalyzer> processedAnalyzers;
+            switch (eventProcessedState.Kind)
+            {
+                case EventProcessedStateKind.Processed:
+                    processedAnalyzers = analysisScope.Analyzers;
+                    break;
+
+                case EventProcessedStateKind.PartiallyProcessed:
+                    processedAnalyzers = eventProcessedState.SubsetProcessedAnalyzers;
+                    break;
+
+                default:
+                    return;
+            }
+
+            await OnEventProcessedCoreAsync(e, processedAnalyzers, analysisStateOpt, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task OnEventProcessedCoreAsync(CompilationEvent e, ImmutableArray<DiagnosticAnalyzer> processedAnalyzers, AnalysisState analysisStateOpt, CancellationToken cancellationToken)
+        {
+            if (analysisStateOpt != null)
+            {
+                await analysisStateOpt.OnCompilationEventProcessedAsync(e, processedAnalyzers, onSymbolAndMembersProcessedAsync).ConfigureAwait(false);
+            }
+            else if (AnalyzerActions.SymbolStartActionsCount > 0 &&
+                e is SymbolDeclaredCompilationEvent symbolDeclaredEvent)
+            {
+                foreach (var analyzer in processedAnalyzers)
+                {
+                    await onSymbolAndMembersProcessedAsync(symbolDeclaredEvent.Symbol, analyzer).ConfigureAwait(false);
+                }
+            }
+
+            async Task onSymbolAndMembersProcessedAsync(ISymbol symbol, DiagnosticAnalyzer analyzer)
+            {
+                if (AnalyzerActions.SymbolStartActionsCount == 0 || symbol.IsImplicitlyDeclared)
+                {
+                    return;
+                }
+
+                _perSymbolAnalyzerActionsCache.TryRemove((symbol, analyzer), out _);
+
+                await processContainerOnMemberCompletedAsync(symbol.ContainingNamespace, symbol, analyzer).ConfigureAwait(false);
+                await processContainerOnMemberCompletedAsync(symbol.ContainingType, symbol, analyzer).ConfigureAwait(false);
+            }
```

Comment:
Should this be a single call with the first argument being `symbol.ContainingType ?? symbol.ContainingNamespace` ?  Symbol can't be a direct member of both.

Reply:
Yes, I will update.

Reply:
Actually, I attempted to change it as per your recommendation and this broke the unit tests for scenario where one or more analyzer has registered symbol start/end actions for Namespace symbols, but none for NamedType symbols. Containing named type of members might already have been processed and marked complete due to no symbol start/end actions, in which case we still want to process the containing namespace for pending symbols that prevent it's end action execution.